docs/configuration/index.md

```diff
@@ -812,7 +812,8 @@ These Coordinator static configurations can be defined in the `coordinator/runti
 |`druid.coordinator.kill.pendingSegments.on`|Boolean flag for whether or not the Coordinator clean up old entries in the `pendingSegments` table of metadata store. If set to true, Coordinator will check the created time of most recently complete task. If it doesn't exist, it finds the created time of the earliest running/pending/waiting tasks. Once the created time is found, then for all dataSources not in the `killPendingSegmentsSkipList` (see [Dynamic configuration](#dynamic-configuration)), Coordinator will ask the Overlord to clean up the entries 1 day or more older than the found created time in the `pendingSegments` table. This will be done periodically based on `druid.coordinator.period.indexingPeriod` specified.|true|
 |`druid.coordinator.kill.on`|Boolean flag for whether or not the Coordinator should submit kill task for unused segments, that is, hard delete them from metadata store and deep storage. If set to true, then for all whitelisted dataSources (or optionally all), Coordinator will submit tasks periodically based on `period` specified. These kill tasks will delete all unused segments except for the last `durationToRetain` period. A whitelist can be set via dynamic configuration `killDataSourceWhitelist` described later.|true|
 |`druid.coordinator.kill.period`|How often to send kill tasks to the indexing service. Value must be greater than `druid.coordinator.period.indexingPeriod`. Only applies if kill is turned on.|P1D (1 Day)|
-|`druid.coordinator.kill.durationToRetain`| Do not kill unused segments in last `durationToRetain`, must be greater or equal to 0. Only applies and MUST be specified if kill is turned on.|`P90D`|
+|`druid.coordinator.kill.durationToRetain`|Only applies if you set `druid.coordinator.kill.on` to `true`. This value is ignored if `druid.coordinator.kill.ignoreDurationToRetain` is `true`. Valid configurations must be a ISO8601 period. Druid will not kill unused segments whose interval end date is beyond `now - durationToRetain`. `durationToRetain` can be a negative ISO8601 period, which would result in `now - durationToRetain` to be in the future.|`P90D`|
```

Comment:
I wonder if we should introduce a concept of last used time which is a timestamp when the segment is used last. The last used time would be a state and can be updated whenever the segment used flag is set from `true` to `false`. The coordinator can use the last used time instead of the end time of the segment to determine whether the segment is eligible for cleanup.

Reply:
@jihoonson I like that idea. The current strategy always seemed odd to me. I see in the code where we mark segments in SqlSegmentsMetadataQuery and that would be easy to change. However, I'm unsure of how we'd safely handle the migration to having a new column in druid_segments table. If you have insight into that piece, I could work on an implementation plan

Reply:
Good question. I would like to avoid adding a new column in `druid_segments` table if possible since it will complicate the migration process. We can perhaps add a new field in `DataSegment` instead which is stored in the `payload` column in the `druid_segments` table. This will increase the cost of `SqlSegmentsMetadataManager.getUnusedSegmentIntervals()` since it will no longer be able to use the index on the end date, but have to read each row, deserialize `payload`, and check the last update time in it. I guess it would be worth to run some benchmark to see how expensive this would be. I guess it will be OK if this ends up a bit expensive as the coordinator doesn't do much work. But, if you think this is going to be too expensive, then I think we have to either add a new column in `druid_segments` table or just stick to your change in this PR. If we are going to add a new column, we will have to document a right upgrade/downgrade path. Providing a script for auto migration would be nice, but not mandatory. 

Another thing we should think about is what the last update time will be for existing unused segments as the new field will be missing for them. Perhaps we can keep the current behavior for the segments that the last update time is missing.

Reply:
Agreed that a schema update is a bit intrusive to the user and should be avoided in all but the most necessary instances. Which this doesn't seem to be since we have multiple known work-a-rounds to a solution including my current one and the one you propose. I'll do a little poking around to see what a payload change would look like and report back. I appreciate the thoughts

Reply:
I think I remember @imply-cheddar mentioning something the other day about adding a field to one of the tables, and building a system to enable the migrations. I don't remember anything more than that, but maybe he can illuminate it for us.

At any rate the "update time" thing could be a different patch.

Reply:
Yes, I'd be 100% okay with this larger change in the approach moving to another patch. I'm definitely interested to hear more of @imply-cheddar thoughts on what the new column would entail. @danprince1 is playing around with ideas regarding a new column in druid_segments as well. His idea is a `last_modified_date` column. However, he is working on a whole separate thread regarding the coordinator polling of used segments, and making that more efficient by only polling what has changed. So the functionality of the column would probably be unrelated to a column regarding "last_used" date that this segment kill stuff could need, but perhaps there is collaboration opportunity on how to handle migrations so there is a standard process. Just a thought.

Reply:
There is an ongoing PR by @AmatyaAvadhanula  that adds a new field to the tasks table. There is going to be code, that runs in a separate thread during startup and does the migration behind the scenes. 
https://github.com/apache/druid/pull/12404/files